### DISCUSSION TITLE
Why not use an already existing computer algebra system?

### DISCUSSION THREAD

**Discussion by Markus40** [clementwzk/OpenCalc#212] [General]
The app used MXparser up until version 1.8.3 which seems to be pretty usable and even has a few advanced functions. Why was it replaced?

Writing a decent algebra system is quite difficult. There are a lot of edge cases and peculiarities even in simple mathematics.

I found lots of bugs in just a few minutes, even without exploiting rounding errors of the underlying 64-bit float system. Why not use an existing computer algebra system, preferably with advanced operations and infinite precision?

**Comment by lefterisgar** [comment]
IMHO you are absolutely right. Avoiding MXparser is like reinventing the wheel, but @Darkempire78 might have removed it due to licensing or performance issues. It's size is around 3 times OpenCalc's size so maybe it added an unwanted layer of complexity to the project.

However MXparser's pros might outweigh the cons because, as you have stated, there are tons of bugs with the current implementation and we need to spend a lot of time fixing those instead of implementing new features or improving UI/UX.

**Comment by clementwzk** [comment]
Hello @Markus40 and thanks for all your feedback, it is extremely important!
OpenCalc doesn't use MXParser anymore since the v2.0.0 for two main reasons:
- The first is that it allows much more freedom and flexibility. MXParser has many known problems, either with accuracy or with features not present. Creating your own implementation allows you to control everything
- The second one is that it's lighter to use its own implementation

Indeed, the current implementation is not yet perfect despite the many betas. However it is only a matter of time. All your bugs will be fixed soon. I invite you to continue your bug searches, it helps us a lot! Thank you also for your suggestions of improvements which will also be included when the settings appear!

**Comment by inson1** [comment]
With #387 we should close this :)
